File: src/main/java/com/nanabell/sponge/nico/storage/PersistenceManager.java
package com.nanabell.sponge.nico.storage;

import org.jetbrains.annotations.NotNull;

import javax.annotation.Nullable;
import java.util.HashMap;
import java.util.Map;

public class PersistenceManager {

    @SuppressWarnings("rawtypes")
    private Map<Class<? extends IdentifiableDaoEnabled<?>>, Persistable> persistableMap = new HashMap<>();

    public <T extends IdentifiableDaoEnabled<T>> void register(@NotNull Persistable<T> persistable) {
        persistableMap.put(persistable.getDaoClass(), persistable);
    }

    @Nullable
    @SuppressWarnings("unchecked")
    public <T extends IdentifiableDaoEnabled<T>> Persistable<T> get(Class<T> clazz) {
        return (Persistable<T>) persistableMap.get(clazz);
    }

    public <T extends IdentifiableDaoEnabled<T>> Persistable<T> getUnchecked(Class<T> clazz) {
        return get(clazz);
    }
}

File: src/main/kotlin/com/nanabell/sponge/nico/storage/PersistenceManager.kt
package com.nanabell.sponge.nico.storage

import java.util.*

class PersistenceManager {
    private val persistableMap: MutableMap<Class<out IdentifiableDaoEnabled<*>>, Persistable<*>> = HashMap()

    fun <T : IdentifiableDaoEnabled<T>> register(persistable: Persistable<T>) {
        persistableMap[persistable.daoClass] = persistable
    }

    @Suppress("UNCHECKED_CAST")
    operator fun <T : IdentifiableDaoEnabled<T>> get(clazz: Class<T>): Persistable<T> {
        return persistableMap[clazz] as Persistable<T>
    }

    fun <T : IdentifiableDaoEnabled<T>> getUnchecked(clazz: Class<T>): Persistable<T> {
        return get(clazz)
    }
}